documentation/tuning_guide.markdown

```diff
@@ -0,0 +1,156 @@
+server for maximum performance and hardware resource utilization.  In this guide,
+we'll highlight some of the most important settings that you can use to get
+the best performance in your environment.
+
+## Puppet Server and JRuby
+
+Before you begin tuning your configuration, it's helpful to have a little bit
+of context on how Puppet Server uses JRuby to handle incoming HTTP requests from
+your Puppet agents.
+
+When Puppet Server starts up, it creates a pool of JRuby interpreters to use
+as workers when we need to execute some of the Puppet Ruby code.  You can think
+of these almost as invidual Ruby "virtual machines" that are controlled by
+Puppet Server; it's not entirely dissimilar to the way that Passenger spawns
+several MRI Ruby processes to hand off work to.
+
+Puppet Server isolates these JRuby instances so that they will only be allowed
+to handle one request at a time.  This ensures that we do not encounter any
+concurrency issues, since the Ruby code is not thread-safe.  When an HTTP request
+comes in to Puppet Server and it determines that some Ruby code will need to be
+executed in order to handle the request, Puppet Server "borrows" a JRuby instance
+from the pool, uses it to do the work, and then "returns" it to the pool.  If
+there are no JRuby instances available in the pool at the time when a request
+comes in (presumably because all of the JRuby instances are already in use handling
+other requests), Puppet Server will block the request until one becomes available.
+
+(In the future, this approach will allow us to do some really powerful things
+such as creating multiple pools of JRubies and isolating each of your Puppet
+environments to a single pool, to ensure that there is no pollution from one
+Puppet environment to the next.)
+
+This brings us to the two most important settings that you can use to tune your
+Puppet Server.
+
+### Number of JRubies
+
+The most important setting that you can use to improve the throughput of your
+Puppet Server installation is the [`max-active-instances`](./configuration.html#puppetserver_conf)
+setting.  The value of this setting is used by Puppet Server to determine how
+many JRuby instances to create when the server starts up.
+
+From a practical perspective, this setting basically controls how many Puppet
+agent runs Puppet Server can handle concurrently.  The minimum value you can
+get away with here is `1`, and if your installation is small enough such that
+you're unlikely to ever have more than one Puppet agent checking in with the
+server at exactly the same time, this will be totally sufficient.
+
+However, if you specify a value of `1` for this setting, and then you have two
+Puppet agent runs hitting the server at the same time, what will happen is that
+the requests being made by the second agent will be effectively blocked until
+the server has finished handling all of the requests from the first agent.
```

Comment:
I intentionally avoiding getting into that level of detail because I thought it would just make things more confusing.  I'm happy to revisit that if @jbondpdx or @nfagerlund suggests that it would be worthwhile.